File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/fabric/mounting/mountitems/MountItemFactory.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.fabric.mounting.mountitems;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.facebook.infer.annotation.Nullsafe;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.fabric.events.EventEmitterWrapper;
import com.facebook.react.uimanager.StateWrapper;

/** Factory class that expose creation of {@link MountItem} */
@Nullsafe(Nullsafe.Mode.LOCAL)
public class MountItemFactory {

  /** @return a {@link DispatchCommandMountItem} for commands identified by an int */
  public static DispatchCommandMountItem createDispatchCommandMountItem(
      int surfaceId, int reactTag, int commandId, @Nullable ReadableArray commandArgs) {
    return new DispatchIntCommandMountItem(surfaceId, reactTag, commandId, commandArgs);
  }

  /** @return a {@link DispatchCommandMountItem} for commands identified by a String */
  public static DispatchCommandMountItem createDispatchCommandMountItem(
      int surfaceId, int reactTag, @NonNull String commandId, @Nullable ReadableArray commandArgs) {
    return new DispatchStringCommandMountItem(surfaceId, reactTag, commandId, commandArgs);
  }

  /** @return a {@link MountItem} that will control the execution of an AccessibilityEvent */
  public static MountItem createSendAccessibilityEventMountItem(
      int surfaceId, int reactTag, int eventType) {
    return new SendAccessibilityEventMountItem(surfaceId, reactTag, eventType);
  }

  /** @return a {@link MountItem} that will be used to preallocate views */
  public static MountItem createPreAllocateViewMountItem(
      int surfaceId,
      int reactTag,
      @NonNull String component,
      @Nullable ReadableMap props,
      @Nullable StateWrapper stateWrapper,
      @Nullable EventEmitterWrapper eventEmitterWrapper,
      boolean isLayoutable) {
    return new PreAllocateViewMountItem(
        surfaceId, reactTag, component, props, stateWrapper, eventEmitterWrapper, isLayoutable);
  }
  /**
   * @return a {@link MountItem} that will be read and execute a collection of MountItems serialized
   *     in the int[] and Object[] received by parameter
   */
  public static MountItem createIntBufferBatchMountItem(
      int surfaceId, int[] intBuf, Object[] objBuf, int commitNumber) {
    return new IntBufferBatchMountItem(surfaceId, intBuf, objBuf, commitNumber);
  }
}

File: packages/react-native/temp/package/ReactAndroid/src/main/java/com/facebook/react/fabric/mounting/mountitems/MountItemFactory.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.fabric.mounting.mountitems

import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.uimanager.StateWrapper

/** Factory class that expose creation of [MountItem] */
public object MountItemFactory {
  /** @return a [DispatchCommandMountItem] for commands identified by an int */
  @JvmStatic
  public fun createDispatchCommandMountItem(
      surfaceId: Int,
      reactTag: Int,
      commandId: Int,
      commandArgs: ReadableArray?
  ): DispatchCommandMountItem =
      DispatchIntCommandMountItem(surfaceId, reactTag, commandId, commandArgs)

  /** @return a [DispatchCommandMountItem] for commands identified by a String */
  @JvmStatic
  public fun createDispatchCommandMountItem(
      surfaceId: Int,
      reactTag: Int,
      commandId: String,
      commandArgs: ReadableArray?
  ): DispatchCommandMountItem =
      DispatchStringCommandMountItem(surfaceId, reactTag, commandId, commandArgs)

  /** @return a [MountItem] that will control the execution of an AccessibilityEvent */
  @JvmStatic
  public fun createSendAccessibilityEventMountItem(
      surfaceId: Int,
      reactTag: Int,
      eventType: Int
  ): MountItem = SendAccessibilityEventMountItem(surfaceId, reactTag, eventType)

  /** @return a [MountItem] that will be used to preallocate views */
  @JvmStatic
  public fun createPreAllocateViewMountItem(
      surfaceId: Int,
      reactTag: Int,
      component: String,
      props: ReadableMap?,
      stateWrapper: StateWrapper?,
      isLayoutable: Boolean
  ): MountItem =
      PreAllocateViewMountItem(surfaceId, reactTag, component, props, stateWrapper, isLayoutable)

  /**
   * @return a [MountItem] that will be read and execute a collection of MountItems serialized in
   *   the int[] and Object[] received by parameter
   */
  @JvmStatic
  public fun createIntBufferBatchMountItem(
      surfaceId: Int,
      intBuf: IntArray,
      objBuf: Array<Any?>,
      commitNumber: Int
  ): MountItem = IntBufferBatchMountItem(surfaceId, intBuf, objBuf, commitNumber)
}
